### DISCUSSION TITLE
The difference between the official and the community Helm Chart in 2022?

### DISCUSSION THREAD

**Discussion by volvocarsmatt** [apache/airflow#26906] [Q&A] [answered]
I just realised I that there are two established Helm Charts for Airflow. One being the official Helm Chart of this repository (`apache-airflow/airflow`) and one is the community Helm Chart in the [airflow-helm repository](https://github.com/airflow-helm/charts) (`airflow-stable/airflow`). Why are there two separate Helm Charts and what are the differences between the two given that airflow is on paper also open source?

The community Helm Chart states that [their project goals](https://github.com/airflow-helm/charts/tree/main/charts/airflow#project-goals) are

> * Ease of Use
> * Great Documentation
> * Support for older Airflow Versions
> * Support for Kubernetes GitOps Tools (like ArgoCD)

[The official Helm Chart](https://airflow.apache.org/docs/helm-chart/stable/index.html) does not even seem to recognise the community Helm Chart's existence, which makes things more confusing. Without being aware of the existence of the two versions, I set up the community Helm Chart because the documentation was incredibly simple to follow for Kubernetes beginners.

A similar [question was asked in 2021](https://github.com/apache/airflow/discussions/14039). The answer there was that the official version was not available and that it was developed by separate teams. [This thread](https://github.com/airflow-helm/charts/issues/211#issuecomment-850296212) also seem to cover the same topic, but I'm not sure if it is accurate anymore.

I guess the code is there for anyone to find out, but does anyone with better Airflow, Kubernetes and Helm Chart expertise know what the practical differences are in 2022?

**Comment by potiuk** [comment, marked as answer]
Creators of Airflow official chart have no influence or impact on the "user community" chart. The user community chart is done by @thesuperzapper  - and it is entirely up to him to maintain and release it. We are communicating and cooperating (for example @thesuperzapper was running Australia's Airflow Summit in 2021 and 2022) and we recognise the value it has for multiple users, but there are some decisions made in the chart that we approached differently.  For some users, the chart from Matthew might be better, for some the official one. The big difference is that if you are corporate users, the official chart is done by the same team that build Airflow and follow the standard release process and are released followint the Apache Software Foundation process- which means they have the same certainty and legal properties that all other ASF projects get - basically the ASF "stamp of approval" that make corporate users reassured that the licence, maintainability, legal properties of the software are kept and guaranteed. You know that you can rely on it being released, maintained, supported no matter what happens to individuals running it. Witth the user community chart it basically depends on @thesuperzapper  being available to release and maintain it, it's not a Foundation-run project. 

BTW. It does not make it "better" or "worse", it's just a statement of fact. 

It's mostly up to the users to choose the one they like. There is no comparision - we do not track and update it, as we are concerned about documenting the ASF "owned" and released code. Maybe @thesuperzapper might want to describe it in his repo (or here), but mostly you should take a look and chose whatever approach you want to take. The "legal" properties and guarantees is the most important diference, I believe you can do pretty similar deployments using any of the charts, but if the "legal" property is not important to you or your employer, I suggest you look at your own use case and see which of the available charts fits better your case.

There are other charts as well - for example Bitnami releases their chart - and there we do not even have any connection to those who create it, this is entirely independent.

We neither want, nor have the capability of doing anythig more. We cannot forbid anyone to build on top of Airflow - this is fundamental foundational value of the Apaache Software Foundation. The ASF releases software and anyone should be able to use it and make a derivative work from this as long as it is marked as "Powered by Apache Airflow" (which we made sure @thesuperzapper 's chart follows). This is absolutely fundamental property of all software released by the foundation.

**Reply by volvocarsmatt** [reply]
A very diplomatic answer :)

I get your point in that this team want to have full ownership of their chart and mostly focus on improving this chart without spending too much time focusing on other projects.

It is true that the code/YAML of both Helm charts is there for anyone to inspect and analyse, but at the moment know enough about  Airflow, Kubernetes and Helm Charts to understand the differences myself. I'll get there eventually, but until then, do you happen to know what "some decisions made in the chart that we approached differently" would include?

What I've found so far myself and in that other thread is this
* The community chart allows, but does not recommend, the hacky way of including more pip packages without setting up docker properly
* The community chart defaults to many version behind Airflow compared with official support (not by design though).
* The community chart has better documentation (not by design)

**Reply by potiuk** [reply]
> A very diplomatic answer :)

It's just well documented state of the case :). Nothing diplomatic in it. It's Radically candid answer :).


> I get your point in that this team want to have full ownership of their chart and mostly focus on improving this chart without spending too much time focusing on other projects.

Just to clarify, this is actually mandated by the Apache Software Foundation rules, not by our "wants". We cannot and should not in any way endorse and describe any software that is not released under the Apache Software Foundation rules. It has legal implications and as a PMC member I have actually legal responsibilty to make sure that the software is run accordingly to Apache Software Foundation policies, and endorsing and "suggesting" that other software might be somehow "ASF" endorsed is something the ASF does not allow us to do. All the software we release and provide as "Apache" software must follow the rules of consensus-driven development, formal voting and releasing the software must be done by the PMC members. So it is not a "want", it's "legal responsibility" of ours. That's why for example we have [Ecosystem](https://airflow.apache.org/ecosystem/) page in our official documentation where some third-party charts are mentioned - but this page has all disclaimers that this is NOT ASF software and it is 3rd-party controlled. This is an absolute maximum we can do there. Comparing our software to other 3rd-party software is not really possible in our official docs. 

But I can of course discuss it here :).

> The community chart allows, but does not recommend, the hacky way of including more pip packages without setting up docker properly

Yes. It's hacky, it causes many problems and this was a major source of differences in opinions and this is one of the main reasons why we have not even attempted to aproach the case where the helm chart of @thesuperzapper will be contributed to Airflow because we could not agree on that (and few other things). This causes many problems for inexperienced users and we prefer to educate them rather than let then burn their own fingers with it. Very, very fresh example here https://github.com/apache/airflow/discussions/26927 - where extra pip packages from Bitnami chart caused some problems we do not even intend to help with because it's strongly agains our vision how it should be done.

> The community chart defaults to many version behind Airflow compared with official support (not by design though).

Indeed we make sure that whenever we release the latest chart, we add features that we added in Airflow which was released since. And also our approach is that Airlfow in it's 2 series (as opposed to end-of-life 1.10) is semver compliant, which means that we make sure that all the 2.* versions at most contain new features, not breaking changes, so we want to educate our users and push them in the direction that they upgrade to latest version of Airflow as soon as humanly possible for them - withoug lagging behind. We do not forbid the user to stay in the old versions and we support it but we strongly recomment to migrate and this is another area when we have different focus (this is not a disagreement of opinion but different focus). Our focus is to push our user to migrate to latest Airflow ASAP. Matthew's focus is to support users who are stick with old versions better and he is much more "fine" with users who stayed behind (it's also part of the business model of the work that Matthew does with supporting his customers - and TBH this is actually pretty cool that somone wants to do it, this makes us much more free to focus on the future of the project rather than look at the past).

> The community chart has better documentation (not by design)

I think it depends on who the users are. We are focused on more "advanced" users who know what they are doing and have experience with charts. We actually even recommend the users who do not know Helm/Kubernetes etc. to install airflow differently - see this documentation where we advise our users to install Airflow in one of the ways that they are most familiar with: https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html and we state what kind of experience they should have.
For us K8S is just one of the ways Airlfow can be installed, and why we [LOVE K8S](https://www.youtube.com/watch?v=Xub3mJVtYpc) we also see it as "one of the ways". Which is also backed by the survey we run https://airflow.apache.org/blog/airflow-survey-2022/#how-do-you-deploy-airflow-multiple-choice where Helm on K8S is big (~40%) but not the only way of deploying Airflow and our chart is just 20%.  This is in a stark contrast with Matthews's approach where Helm chart is the only thing Matthew dos. 

Of course improving the docs is most welcome - Airlfow is created by > 2200 contributors and if you think some of the docs can be improved - you are absolutely free to help with that and make PRs that can improve it. This is **super easy** - just go to the docs, click "Suggest a change in this page" button at the bottom-right and you will open a Pull Request where (without any tools or setup) you will be able to update the docs. I very warmly welcome you to do so (after all the explanations you get, you have much better context for it)  :)
